crates/wasi/src/preview2/pipe.rs

```diff
@@ -7,227 +7,835 @@
+                Ok(Ok(StreamState::Open)) => {
+                    self.state = Some(WriteState::Ready);
+                    self.send(bytes)
+                }
+
+                Ok(Ok(StreamState::Closed)) => {
+                    self.state = None;
+                    Ok((0, StreamState::Closed))
+                }
+
+                Ok(Err(e)) => {
+                    self.state = None;
+                    Err(e.into())
+                }
 
-    /// Try to convert this `WritePipe<W>` back to the underlying `W` type.
-    ///
-    /// This will fail with `Err(self)` if multiple references to the underlying `W` exist.
-    pub fn try_into_inner(mut self) -> Result<W, Self> {
-        match Arc::try_unwrap(self.writer) {
-            Ok(rc) => Ok(RwLock::into_inner(rc).unwrap()),
-            Err(writer) => {
-                self.writer = writer;
-                Err(self)
+                Err(TryRecvError::Empty) => {
+                    self.state = Some(WriteState::Pending);
+                    Ok((0, StreamState::Open))
+                }
+
+                Err(TryRecvError::Disconnected) => {
+                    unreachable!("task shouldn't die while pending")
+                }
+            },
+            Some(WriteState::Err(_)) => {
+                // Move the error payload out of self.state, because errors are not Copy,
+                // and set self.state to None, because the stream is now closed.
+                if let Some(WriteState::Err(e)) = self.state.take() {
+                    Err(e.into())
+                } else {
+                    unreachable!("self.state shown to be Some(Err(e)) in match clause")
+                }
             }
+
+            None => Ok((0, StreamState::Closed)),
         }
     }
 
-    fn borrow(&self) -> std::sync::RwLockWriteGuard<W> {
-        RwLock::write(&self.writer).unwrap()
+    async fn ready(&mut self) -> Result<(), Error> {
+        match &self.state {
+            Some(WriteState::Pending) => match self.result_receiver.recv().await {
+                Some(Ok(StreamState::Open)) => {
+                    self.state = Some(WriteState::Ready);
+                }
+
+                Some(Ok(StreamState::Closed)) => {
+                    self.state = None;
+                }
+
+                Some(Err(e)) => {
+                    self.state = Some(WriteState::Err(e));
+                }
+
+                None => unreachable!("task shouldn't die while pending"),
+            },
+
+            Some(WriteState::Ready | WriteState::Err(_)) | None => {}
+        }
+
+        Ok(())
     }
 }
 
-impl WritePipe<io::Cursor<Vec<u8>>> {
-    /// Create a new writable virtual pipe backed by a `Vec<u8>` buffer.
-    pub fn new_in_memory() -> Self {
-        Self::new(io::Cursor::new(vec![]))
+/// A stream that is never able to provide input. It stays open forever, but is never ready for
+/// reading.
+pub struct EmptyInputStream;
```

Comment:
I find the semantics of this type a bit dubious, but is this necessary? Would it be possible to remove this?

Reply:
yeah, we did it because it corresponded to one of the output streams but now that you point it out, its really not necessary at all